src/hdmf/container.py

```diff
@@ -839,3 +842,297 @@ def __build_conf_methods(cls, conf_dict, conf_index, multi):
         get = conf_dict.get('get')
         if get is not None:
             setattr(cls, get, cls.__make_get(get, attr, container_type))
+
+
+class Row(object, metaclass=ExtenderMeta):
+    """
+    A class for representing rows from a Table.
+
+    The Table class can be indicated with the __table__. Doing so
+    will set constructor arguments for the Row class and ensure that
+    Row.idx is set appropriately when a Row is added to the Table. It will
+    also add functionality to the Table class for getting Row objects.
+
+    Note, the Row class is not needed for working with Table objects. This
+    is merely convenience functionality for working with Tables.
+    """
+
+    __table__ = None
+
+    @property
+    def idx(self):
+        """The index of this row in its respective Table"""
+        return self.__idx
+
+    @idx.setter
+    def idx(self, val):
+        if self.__idx is None:
+            self.__idx = val
+        else:
+            raise ValueError("cannot reset the ID of a row object")
+
+    @property
+    def table(self):
+        """The Table this Row comes from"""
+        return self.__table
+
+    @table.setter
+    def table(self, val):
+        if val is not None:
+            self.__table = val
+        if self.idx is None:
+            self.idx = self.__table.add_row(**self.todict())
+
+    @ExtenderMeta.pre_init
+    def __build_row_class(cls, name, bases, classdict):
+        table_cls = getattr(cls, '__table__', None)
+        if table_cls is not None:
+            columns = getattr(table_cls, '__columns__')
+            if cls.__init__ == bases[-1].__init__:  # check if __init__ is overridden
+                columns = deepcopy(columns)
+                func_args = list()
+                for col in columns:
+                    func_args.append(col)
+                func_args.append({'name': 'table', 'type': Table, 'default': None,
+                                  'help': 'the table this row is from'})
+                func_args.append({'name': 'idx', 'type': int, 'default': None,
+                                  'help': 'the index for this row'})
+
+                @docval(*func_args)
+                def __init__(self, **kwargs):
```

Comment:
I think there should be a base `Row.__init__` that initializes `self.__keys`, `self.__idx`, `self.__table`, and `self.__todict()` and the Row subclass should call that init. Otherwise if line 891 is false, several methods will break.

Reply:
I'd rather just raise an error if `__table__` is not set.